core/src/main/scala/kafka/server/FetchSession.scala

```diff
@@ -0,0 +1,720 @@
+      })
+      cachedPartitions
+    }
+    val responseSessionId = cache.maybeCreateSession(time.milliseconds(), isFromFollower,
+        updates.size(), createNewSession)
+    debug("Full fetch context with session id %d returning %s".format(
+      responseSessionId, partitionsToLogString(updates.keySet())))
+    new FetchResponse(Errors.NONE, updates, 0, responseSessionId)
+  }
+}
+
+/**
+  * The fetch context for an incremental fetch request.
+  *
+  * @param time         The clock to use.
+  * @param reqMetadata  The request metadata.
+  * @param session      The incremental fetch request session.
+  */
+class IncrementalFetchContext(private val time: Time,
+                              private val reqMetadata: JFetchMetadata,
+                              private val session: FetchSession) extends FetchContext {
+
+  override def getFetchOffset(tp: TopicPartition): Option[Long] = session.getFetchOffset(tp)
+
+  override def foreachPartition(fun: (TopicPartition, FetchRequest.PartitionData) => Unit): Unit = {
+    // Take the session lock and iterate over all the cached partitions.
+    session.synchronized {
+      session.partitionMap.iterator().asScala.foreach(part => {
+        fun(new TopicPartition(part.topic, part.partition), part.reqData())
+      })
+    }
+  }
+
+  override def updateAndGenerateResponseData(updates: FetchSession.RESP_MAP): FetchResponse = {
+    session.synchronized {
+      // Check to make sure that the session epoch didn't change in between
+      // creating this fetch context and generating this response.
+      val expectedEpoch = JFetchMetadata.nextEpoch(reqMetadata.epoch())
+      if (session.epoch != expectedEpoch) {
+        info("Incremental fetch session %d expected epoch %d, but got %d.  " +
+          "Possible duplicate request.".format(session.id, expectedEpoch, session.epoch))
+        new FetchResponse(Errors.INVALID_FETCH_SESSION_EPOCH, new FetchSession.RESP_MAP, 0, session.id)
+      } else {
+        // Iterate over the update list.  Prune updates which don't need to be sent.
+        val iter = updates.entrySet().iterator()
+        while (iter.hasNext()) {
+          val entry = iter.next()
+          val topicPart = entry.getKey
+          val respData = entry.getValue
+          val cachedPart = session.partitionMap.find(new CachedPartition(topicPart))
+          val mustRespond = cachedPart.updateResponseData(respData)
+          if (mustRespond) {
+            // Move this to the end of the cached partition map.
+            // This is important for ensuring fairness when lots of partitions
+            // have data to return.
+            session.partitionMap.remove(cachedPart)
+            session.partitionMap.mustAdd(cachedPart)
+          } else {
+            // Do not include this partition in the FetchResponse.
+            iter.remove()
+          }
+        }
+        debug("Incremental fetch context with session id %d returning %s".format(
```

Comment:
On the server side, we have moved to the s convention for building a string, instead of format.

Reply:
OK.  I will change it over to the 's' convention